src/Microsoft.Identity.Client/Instance/Discovery/InstanceDiscoveryManager.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.TelemetryCore;
+
+namespace Microsoft.Identity.Client.Instance.Discovery
+{
+    internal class InstanceDiscoveryManager : IInstanceDiscoveryManager
+    {
+        // TODO: move caching to a different class
+        private static readonly ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry> s_cache =
+            new ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry>();
+
+        private readonly ICoreLogger _logger; // TODO: needed?
+        private readonly IHttpManager _httpManager;
+        private readonly ITelemetryManager _telemetryManager;
+
+
+        public InstanceDiscoveryManager(ICoreLogger logger, IHttpManager httpManager, ITelemetryManager telemetryManager, bool shouldClearCaches)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _httpManager = httpManager ?? throw new ArgumentNullException(nameof(httpManager));
+            _telemetryManager = telemetryManager ?? throw new ArgumentNullException(nameof(telemetryManager));
+
+            if (shouldClearCaches)
+            {
+                s_cache.Clear();
+            }
+        }
+
+        public async Task<InstanceDiscoveryMetadataEntry> GetMetadataEntryAsync(Uri authority, RequestContext requestContext)
```

Comment:
Most of rest of the logic in this file was copied across from the old AadInstanceDIscovery